chia/plot_sync/sender.py

```diff
@@ -0,0 +1,303 @@
+    sync_id: uint64
+    message_type: ProtocolMessageTypes
+    message_id: uint64
+    payload_type: Any
+    args: Any
+
+    def generate(self) -> Tuple[PlotSyncIdentifier, Any]:
+        identifier = PlotSyncIdentifier(uint64(int(time.time())), self.sync_id, self.message_id)
+        payload = self.payload_type(identifier, *self.args)
+        return identifier, payload
+
+
+@dataclass
+class ExpectedResponse:
+    message_type: ProtocolMessageTypes
+    identifier: PlotSyncIdentifier
+    message: Optional[PlotSyncResponse] = None
+
+    def __str__(self) -> str:
+        return (
+            f"expected_message_type: {self.message_type.name}, "
+            f"expected_identifier: {self.identifier}, message {self.message}"
+        )
+
+
+class Sender:
+    _plot_manager: PlotManager
+    _connection: Optional[WSChiaConnection]
+    _sync_id: uint64
+    _next_message_id: uint64
+    _messages: List[MessageGenerator]
+    _last_sync_id: uint64
+    _stop_requested = False
+    _task: Optional[asyncio.Task]
+    _lock: threading.Lock
+    _response: Optional[ExpectedResponse]
+
+    def __init__(self, plot_manager: PlotManager) -> None:
+        self._plot_manager = plot_manager
+        self._connection = None
+        self._sync_id = uint64(0)
+        self._next_message_id = uint64(0)
+        self._messages = []
+        self._last_sync_id = uint64(0)
+        self._stop_requested = False
+        self._task = None
+        self._lock = threading.Lock()
+        self._response = None
+
+    def __str__(self) -> str:
+        return f"sync_id {self._sync_id}, next_message_id {self._next_message_id}, messages {len(self._messages)}"
+
+    async def start(self) -> None:
+        if self._task is not None and self._stop_requested:
+            await self.await_closed()
+        if self._task is None:
+            self._task = asyncio.create_task(self._run())
+            if not self._plot_manager.initial_refresh() or self._sync_id != 0:
+                self._reset()
+        else:
+            raise AlreadyStartedError()
+
+    def stop(self) -> None:
+        self._stop_requested = True
+
+    async def await_closed(self) -> None:
+        if self._task is not None:
+            await self._task
+        self._task = None
+        self._reset()
+        self._stop_requested = False
+
+    def set_connection(self, connection: WSChiaConnection) -> None:
+        if connection.connection_type != NodeType.FARMER:
+            raise InvalidConnectionTypeError(connection.connection_type, NodeType.HARVESTER)
+        self._connection = connection
+
+    def bump_next_message_id(self) -> None:
+        self._next_message_id = uint64(self._next_message_id + 1)
+
+    def _reset(self) -> None:
+        log.debug(f"_reset {self}")
+        self._last_sync_id = uint64(0)
+        self._sync_id = uint64(0)
+        self._next_message_id = uint64(0)
+        self._messages.clear()
+        if self._lock.locked():
+            self._lock.release()
```

Comment:
I think it's being used as a boolean , although I think a more common pattern might be to acquire the lock and set a boolean `with self._lock: is_syncing = True|False` as needed?

Reply:
It _could_ happen that the `PlotManager` runs the next refresh interval before the current sync event is done and the lock lets the `PlotManager` thread which calls `Harvester._plot_refresh_callback` wait in `Sender.sync_start` before it triggers the next plot sync start event until the current one is done. From my understanding with your idea we would need to lock the lock, check if `is_syncing == False` and sleep otherwise and wait the the `is_syncing` to become `False`. So we would re-implement the locking?

Reply:
Yes, that doesn't seem any better. An event perhaps (set,clear,isset,wait) - but it seems that more or less exactly the code you have, so I'm not sure it does anything.